### PR TITLE
working trade window test

### PR DESCRIPTION
**Pull request by Mathis-Paul-62**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -6,6 +6,8 @@ import update_chunks from './chunk/update.js'
 import EventEmitter from 'events'
 import { position_change_event } from './events.js'
 import { chunk_change_event, chunk_position } from './chunk.js'
+import { openTrade } from './trade/trade.js'
```
Comment (DeltaEvo) [inline src/index.js:9]: change the casing to lowercase `open_trade`

File: src/index.js [hunk 2]
```diff
@@ -6,6 +6,8 @@ import update_chunks from './chunk/update.js'
...
+import { spawnVillager } from './trade/spawnVillager.js'
```
Comment (DeltaEvo) [inline src/index.js:10]: change the casing to lowercase `spawn_villager`

File: src/index.js [hunk 3]
```diff
@@ -21,11 +23,16 @@ server.on('login', (client) => {
     events: new EventEmitter(),
     gameMode: 1,
   }
-
+  client.on('end', console.error)
```
Comment (DeltaEvo) [inline src/index.js:26]: Remove

File: src/index.js [hunk 4]
```diff
@@ -21,11 +23,16 @@ server.on('login', (client) => {
...
   login(state)
   position_change_event(state)
   chunk_change_event(state)
   update_chunks(state)
+  spawnVillager(state)
+  openTrade(state)
+  
+  client.on('error', console.log)
```
Comment (DeltaEvo) [inline src/index.js:34]: Remove

File: src/trade/spawnVillager.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import UUID from 'uuid-1345'
+
+export function spawnVillager({client}) {
+    let mob = {
```
Comment (DeltaEvo) [inline src/trade/spawnVillager.js:4]: const

File: src/trade/spawnVillager.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        type: 93, // villager
+        x: 469,
+        y: 163, // in front of the spawn
+        z: 649,
+        yaw: 0,
+        pitch: 0,
+        headPitch: 0,
+        velocityX: 400,
+        velocityY: 400,
+        velocityZ: 400
```
Comment (DeltaEvo) [inline src/trade/spawnVillager.js:16]: Why 400 ?

File: src/trade/trade.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
+export function openTrade({ client }) {
+
+  client.on("use_entity", (evt) => {
+    
+    if(evt.target=='69420' && evt.mouse==2 && evt.sneaking==false){
```
Comment (DeltaEvo) [inline src/trade/trade.js:5]: Use strict equality https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality

File: src/trade/trade.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      
+      const trade = {
+        windowId: 1,
+        trades: [
+          {
+            inputItem1: {
+              present: true,
+              itemId: 581,            //https://minecraft-data.prismarine.js.org/?v=1.16&d=items
```
Comment (DeltaEvo) [inline src/trade/trade.js:13]: Use https://github.com/PrismarineJS/node-minecraft-data/blob/master/doc/api.md#minecraft-dataitemsbyname

File: src/trade/trade.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+              itemCount: 1,
+            },
+            outputItem: {
+              present: true,          //https://wiki.vg/Protocol#Trade_List
```
Comment (DeltaEvo) [inline src/trade/trade.js:17]: Remove comment

File: src/trade/trade.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+              itemId: 595,            //https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.16/protocol.json#L2264-L2321
```
Comment (DeltaEvo) [inline src/trade/trade.js:18]: Use https://github.com/PrismarineJS/node-minecraft-data/blob/master/doc/api.md#minecraft-dataitemsbyname

File: src/trade/trade.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+              itemCount: 1,
+            },
+            inputItem2: {
+              present: true,
+              itemId: 581,
```
Comment (DeltaEvo) [inline src/trade/trade.js:23]: Use https://github.com/PrismarineJS/node-minecraft-data/blob/master/doc/api.md#minecraft-dataitemsbyname

File: src/trade/trade.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+              itemCount: 1,
+            },
+            maximumNbTradeUses: 99999, //mandatory
```
Comment (DeltaEvo) [inline src/trade/trade.js:26]: Can we not set something like 0 or -1 ?

File: src/trade/trade.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+          }
+          
+        ]
+      }
+
+      const window = {
+        windowId: 1,
+        inventoryType: 18, 
+        windowTitle: '{"text": "trade"}',
```
Comment (DeltaEvo) [inline src/trade/trade.js:35]: Use JSON.stringify

File: src/trade/trade.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+            },
+            maximumNbTradeUses: 99999, //mandatory
+          }
+          
+        ]
+      }
+
+      const window = {
+        windowId: 1,
+        inventoryType: 18,
```
Comment (DeltaEvo) [inline src/trade/trade.js:34]: Copy this part from prismarine-windows
https://github.com/PrismarineJS/prismarine-windows/blob/master/index.js#L8-L33
into a `inventories` variable
And use it like
```js
inventoryType: inventories["minecraft:merchant"].type
```

File: src/trade/trade.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
+export function openTrade({ client }) {
+
+  client.on("use_entity", (evt) => {
```
Comment (DeltaEvo) [inline src/trade/trade.js:3]: Destructure evt Object http://es6-features.org/#ParameterContextMatching

File: src/trade/spawnVillager.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
+import UUID from 'uuid-1345'
+
+export function spawnVillager({client}) {
+    let mob = {
+        entityId: 69420, // temporaire 
+        entityUUID: UUID.v4(),
+        type: 93, // villager
```
Comment (DeltaEvo) [inline src/trade/spawnVillager.js:7]: Use https://github.com/PrismarineJS/node-minecraft-data/blob/master/doc/api.md#minecraft-dataentitiesbyname

**Review by DeltaEvo — CHANGES_REQUESTED** [review]

File: src/trade/trade.js (hunk 12 above)
Comment (Mathis-Paul-62) [reply src/trade/trade.js:26]: 0 or -1 doesn't work 
but there may be an alternative, i'll look into it later.

File: src/index.js (hunk 1 above)
Comment (Sceat) [reply src/index.js:9]: to snakecase*

File: src/index.js (hunk 3 above)
Comment (Sceat) [reply src/index.js:26]: please provide better description on your reviews

File: src/trade/spawnVillager.js (hunk 5 above)
Comment (Sceat) [reply src/trade/spawnVillager.js:4]: you should explain why using a `const` is better than a `let`, provide some example and documentation

File: src/trade/trade.js (hunk 15 above)
Comment (Sceat) [reply src/trade/trade.js:3]: or do not use parenthesis

File: src/trade/trade.js (hunk 7 above)
Comment (Sceat) [reply src/trade/trade.js:5]: `==`

image

File: src/trade/trade.js (hunk 9 above)
Comment (Sceat) [reply src/trade/trade.js:17]: david you sound like a bot

**Review by Sceat — APPROVED** [review]

File: src/trade/spawnVillager.js (hunk 6 above)
Comment (Mathis-Paul-62) [reply src/trade/spawnVillager.js:20]: default speed of a mob if he moves (forgot to remove it since it's a merchant)

File: src/index.js (hunk 1 above)
Comment (Mathis-Paul-62) [reply src/index.js]: done

File: src/index.js (hunk 3 above)
Comment (Mathis-Paul-62) [reply src/index.js:26]: useless logs removed

**Review by DeltaEvo — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 17]
```diff
@@ -21,11 +23,12 @@ server.on('login', (client) => {
     events: new EventEmitter(),
     gameMode: 1,
   }
-
```
Comment (DeltaEvo) [inline src/index.js:24]: Can you re add this line :)

**Review by DeltaEvo — APPROVED** [review]

**Pull request merged**